### DISCUSSION TITLE
Using ESM adds considerable load time

### DISCUSSION THREAD

**Discussion by brenc** [josdejong/mathjs#3007] [Q&A] [answered]
I thought I'd bring this up here before opening a ticket on this just in case I'm actually doing something wrong here. Recently I've been working on moving a project over to ESM. One thing I noticed was that loading the `round` function like this:

`import { round } from 'mathjs'`

Adds about ten seconds to the load time of my application. I have narrowed it down to the above import statement. If I import like so:

`import { round } from 'mathjs/number'`

The app takes about 6 seconds to load. If I isolate this code to a test file:
```
import { round } from 'mathjs'
console.log(round(0.5))

$ time node test_mathjs.js 
1

real    0m1.119s
user    0m1.506s
sys     0m0.302s


import { round } from 'mathjs/number'
console.log(round(0.5))

$ time node test_mathjs.js 
1

real    0m0.661s
user    0m0.848s
sys     0m0.114s
```
 
The timing here tracks but for some reason it takes ten times as long for my app to start. I used the `blocked` module and have a few stack traces. Here is one:

```
[DEBUG] brenbot: event loop blocked for 137.6751337890625ms, operation started here: [
  '    at AsyncHook.init (/app/node_modules/blocked-at/index.js:31:11)',
  '    at emitInitNative (node:internal/async_hooks:202:43)',
  '    at emitInitScript (node:internal/async_hooks:505:3)',
  '    at promiseInitHook (node:internal/async_hooks:324:3)',
  '    at promiseInitHookWithDestroyTracking (node:internal/async_hooks:328:3)',
  '    at link (node:internal/modules/esm/module_job:67:32)',
  '    at new ModuleJob (node:internal/modules/esm/module_job:88:19)',
  '    at ESMLoader.#createModuleJob (node:internal/modules/esm/loader:480:17)',
  '    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:434:34)',
  '    at async ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:78:21)',
  '    at async Promise.all (index 7)',
  '    at async link (node:internal/modules/esm/module_job:83:9)'
]
```

Definitely looks like this is as a result of the ESM loader. The same thing happens when I use lodash, however, I can import specific functions like this:

`import has from 'lodash-es/has.js'`

Which almost completely eliminates the loading delay. I've tried this in various ways with mathjs but it doesn't look like it's possible?

Side note: moving this project to ESM has been one of the most painful development experiences I've had in years. Can only imagine what a nightmare converting a larger project with multiple developers would be. My personal policy now is that I will NOT convert any more of my projects unless absolutely necessary. Only new projects will be using ESM. I'm glad that mathjs continues to support CJS because I use it in several places.

**Comment by kungfooman** [comment]
I don't know how you collect 10 extra seconds, but ESM is definitely slower for me too, issue: https://github.com/nodejs/node/issues/44186

**test_math.cjs**
```js
const { round } = require('mathjs');
```

**test_math.mjs**
```js
import { round } from 'mathjs'
```

```
(base) homepc@home-pc:/var/www/html/mathjs/test$ time node test_mathjs.cjs 

real	0m0,467s
user	0m0,677s
sys	0m0,126s
(base) homepc@home-pc:/var/www/html/mathjs/test$ time node test_mathjs.mjs 

real	0m0,740s
user	0m1,149s
sys	0m0,068s
```

Both still relatively fast, most time is spent in the file system, e.g. loading the browser bundle:

```js
const { round } = require('./node_modules/mathjs/lib/browser/math.js');
```

```js
real	0m0,116s
user	0m0,100s
sys	0m0,021s
```

Which node version, which OS? I'm on Linux Mint and `node -v` is `v18.12.1`

**Comment by brenc** [comment]
Node v18.17.1, Docker on Debian 11 x64, Proxmox VM on enterprise servers with AMD EPYC 7371 procs and NVMe drives.

I wrote a small test app to more clearly illustrate the difference in load times:

https://github.com/brenc/testmathjs

On my dev VM I'm seeing lodash load about 15x faster than mathjs. Tested on my Macbook M1 Pro as well and am seeing the same thing. Here is a typical result on my dev VM:
```
mathjs load time: 1250ms, lodash load time: 83ms, diff: 1167ms, lodash loaded 15x faster than mathjs
```
And on my Macbook Pro:
```
mathjs load time: 741ms, lodash load time: 45ms, diff: 696ms, lodash loaded 16x faster than mathjs
```
Not sure why this leads to 10s+ load times in my app but I ended up just switching this one app to the lodash provided round function for now.

**Comment by josdejong** [comment, marked as answer]
Just using mathjs for the `round` function looks like overkill, if a more lightweight library like lodash works better just use that :).

I've indeed also noticed that ESM is much slower than CJS. We cannot do anything about that in this library though.

mathjs is quite a large library, and by default it loads the whole library, which is much if you only use a single function. The ESM and CJS contains hundreds of files, since every function is written in a separate file. Loading separate files is relatively slow, and for example loading the UMD bundle of mathjs is much faster. We are making plans to make mathjs more modular, but that will take time.
